File: ovikl-android/app/src/main/java/com/caoutch/transnet/TripSingleton.java
package com.caoutch.transnet;

import android.util.Log;

import com.caoutch.transnet.database.Trip;

import org.json.JSONException;
import org.json.JSONObject;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;
import java.util.TimeZone;

/**
 * Created by caoutch on 4/21/2018.
 */

public class TripSingleton {
    private SimpleDateFormat dateFormatter=new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'");

    private transient static TripSingleton mInstance;
    private String TAG="TripSingleton";


    public String _id;
    public Date createTime;
    public Date updateTime;
    //public Date startTime;
    //public Date endTime;
    public String state;
    public String clientId;
    public String driverId;
    public Double clientLat;
    public Double clientLng;
    public Double driverLat;
    public Double driverLng;
    public Double prMin;
    public Double prBase;
    public Double prKM;
    public Double prMinute;
    //public Double prLngKM;
    //public Double prLngMinute;
    //public Double lngKM;
    public String cancelledBy;
    public String cur;
    public Double distance;
    public Double duration;
    public Double cost;
    public String msgId;


    private String lang;

    private TripSingleton(){
    }
    public static synchronized TripSingleton getInstance() {
        if (mInstance == null) {
            mInstance = new TripSingleton();

            Log.i("TripSingleton", "new instance");

        }
        else
            Log.i("TripSingleton", "old instance");
        return mInstance;
    }

    public JSONObject getMap(){
        JSONObject map = new JSONObject();
        try {
            map.put("_id",_id);
            /*if(createTime!=null)
                map.put("createTime",dateFormatter.format(createTime));
            if(updateTime!=null)
                map.put("updateTime",dateFormatter.format(updateTime));
            if(startTime!=null)
                map.put("startTime",dateFormatter.format(startTime));
            if(endTime!=null)
                map.put("endTime",dateFormatter.format(endTime));*/
            if(state!=null)
                map.put("state",state);
            if(clientId!=null)
                map.put("clientId",clientId);
            if(driverId!=null)
                map.put("driverId",driverId);
            if(clientLat!=null)
                map.put("clientLat",clientLat.toString());
            if(clientLng!=null)
                map.put("clientLng",clientLng.toString());
            if(driverLat!=null)
                map.put("driverLat",driverLat.toString());
            if(driverLng!=null)
                map.put("driverLng",driverLng.toString());
            /*if(prMin!=null)
                map.put("prMin",prMin.toString());
            if(prBase!=null)
                map.put("prBase",prBase.toString());
            if(prKM!=null)
                map.put("prKM",prKM.toString());
            if(prMinute!=null)
                map.put("prMinute",prMinute.toString());
            if(prLngKM!=null)
                map.put("prLngKM",prLngKM.toString());
            if(prLngMinute!=null)
                map.put("prLngMinute",prLngMinute.toString());
            if(lngKM!=null)
                map.put("lngKM",lngKM.toString());*/
            if(cur!=null)
                map.put("cur",cur);
            if(cancelledBy!=null)
                map.put("cancelledBy",cancelledBy);
            if(distance!=null)
                map.put("distance",distance);
            if(duration!=null)
                map.put("duration",duration);
            if(cost!=null)
                map.put("cost",cost);

        } catch (JSONException e) {
            e.printStackTrace();
        }
        return map;
    }



    public void setJson(JSONObject tripJson){
        try {
            if(lang==null)
                lang= Locale.getDefault().getLanguage();
            dateFormatter.setTimeZone(TimeZone.getTimeZone("GMT"));
            if(tripJson.has("msgId"))
                msgId=tripJson.getString("msgId");
            if(tripJson.has("_id"))
                _id=tripJson.getString("_id");
            if(tripJson.has("createTime"))
                createTime=dateFormatter.parse(tripJson.getString("createTime"));
            if(tripJson.has("updateTime"))
                updateTime=dateFormatter.parse(tripJson.getString("updateTime"));
            /*if(tripJson.has("startTime"))
                startTime=dateFormatter.parse(tripJson.getString("startTime"));
            if(tripJson.has("endTime"))
                endTime=dateFormatter.parse(tripJson.getString("endTime"));*/
            if(tripJson.has("state"))
                state=tripJson.getString("state");
            if(tripJson.has("clientId"))
                clientId=tripJson.getString("clientId");
            if(tripJson.has("driverId"))
                driverId=tripJson.getString("driverId");
            if(tripJson.has("clientLat"))
                clientLat=tripJson.getDouble("clientLat");
            if(tripJson.has("clientLng"))
                clientLng=tripJson.getDouble("clientLng");
            if(tripJson.has("driverLat"))
                driverLat=tripJson.getDouble("driverLat");
            if(tripJson.has("driverLng"))
                driverLng=tripJson.getDouble("driverLng");
            if(tripJson.has("prMin"))
                prMin=tripJson.getDouble("prMin");
            if(tripJson.has("prBase"))
                prBase=tripJson.getDouble("prBase");
            if(tripJson.has("prKM"))
                prKM=tripJson.getDouble("prKM");
            if(tripJson.has("prMinute"))
                prMinute=tripJson.getDouble("prMinute");
            /*if(tripJson.has("prLngKM"))
                prLngKM=tripJson.getDouble("prLngKM");
            if(tripJson.has("prLngMinute"))
                prLngMinute=tripJson.getDouble("prLngMinute");
            if(tripJson.has("lngKM"))
                lngKM=tripJson.getDouble("lngKM");*/
            if(tripJson.has(lang+"cur"))
                cur=tripJson.getString(lang+"cur");
            else if(tripJson.has("cur"))
                cur=tripJson.getString("cur");

            if(tripJson.has("cancelledBy"))
                cancelledBy=tripJson.getString("cancelledBy");
            if(tripJson.has("distance"))
                distance=tripJson.getDouble("distance");
            if(tripJson.has("duration"))
                duration=tripJson.getDouble("duration");
            if(tripJson.has("cost"))
                cost=tripJson.getDouble("cost");

        } catch (JSONException e) {
            Log.e("",e.getMessage());
        } catch (ParseException e) {
            Log.e("",e.getMessage());
        }
    }

    public void reset(){
        if (mInstance != null) {
            mInstance._id= null;
            mInstance.createTime= null;
            mInstance.updateTime= null;
            //mInstance.startTime= null;
            //mInstance.endTime= null;
            mInstance.state= null;
            mInstance.clientId= null;
            mInstance.driverId= null;
            mInstance.clientLat= null;
            mInstance.clientLng= null;
            mInstance.driverLat= null;
            mInstance.driverLng= null;
            mInstance.prMin= null;
            mInstance.prBase= null;
            mInstance.prKM= null;
            mInstance.prMinute= null;
            /*mInstance.prLngKM= null;
            mInstance.prLngMinute= null;
            mInstance.lngKM= null;*/
            mInstance.cur= null;
            mInstance.cancelledBy= null;
            mInstance.distance= null;
            mInstance.duration= null;
            mInstance.cost= null;
            mInstance.msgId= null;

        }
    }

    public Trip getTrip(){

        Trip t = new Trip();
        t._id= _id;
        t.createTime= createTime;
        t.updateTime= updateTime;
        //t.startTime= startTime;
        //t.endTime= endTime;
        t.state= state;
        t.clientId= clientId;
        t.driverId= driverId;
        t.clientLat= clientLat;
        t.clientLng= clientLng;
        t.driverLat= driverLat;
        t.driverLng= driverLng;
        t.prMin= prMin;
        t.prBase= prBase;
        t.prKM= prKM;
        t.prMinute= prMinute;
        /*t.prLngKM= prLngKM;
        t.prLngMinute= prLngMinute;
        t.lngKM= lngKM;*/
        t.cur= cur;
        t.cancelledBy= cancelledBy;
        t.distance= distance;
        t.duration=duration;
        t.cost=cost;
        t.msgId=msgId;

        return t;
    }


    public void setSingleton(Trip trip){
        if(trip._id!=null)
            _id= trip._id;
        if(trip.createTime!=null)
            createTime= trip.createTime;
        if(trip.updateTime!=null)
            updateTime= trip.updateTime;
        //t.startTime= startTime;
        //t.endTime= endTime;
        if(trip.state!=null)
            state= trip.state;
        if(trip.clientId!=null)
            clientId= trip.clientId;
        if(trip.driverId!=null)
            driverId= trip.driverId;
        if(trip.clientLat!=null)
            clientLat= trip.clientLat;
        if(trip.clientLng!=null)
            clientLng= trip.clientLng;
        if(trip.driverLat!=null)
            driverLat= trip.driverLat;
        if(trip.driverLng!=null)
            driverLng= trip.driverLng;
        if(trip.prMin!=null)
            prMin= trip.prMin;
        if(trip.prBase!=null)
            prBase= trip.prBase;
        if(trip.prKM!=null)
            prKM= trip.prKM;
        if(trip.prMinute!=null)
            prMinute= trip.prMinute;
        /*prLngKM= trip.prLngKM;
        prLngMinute= trip.prLngMinute;
        lngKM= trip.lngKM;*/
        if(trip.cur!=null)
            cur= trip.cur;
        if(trip.cancelledBy!=null)
            cancelledBy= trip.cancelledBy;
        if(trip.distance!=null)
            distance= trip.distance;
        if(trip.distance!=null)
            distance=trip.distance;
        if(trip.cost!=null)
            cost=trip.cost;
        if(trip.msgId!=null)
            msgId=trip.msgId;
    }

    /*public void save(final AppDatabase database){
        AsyncTask.execute(new Runnable() {
            @Override
            public void run() {
                TripMessage message = new TripMessage();
                message.tripId = _id;
                message.state = state;

                message.message = getString(R.string.driver_cancel_trip);

                message.senderId=driverId;
                message.msgId = msgId;
                //message._id = _id;
                message.createTime = updateTime;
                database.tripMessageDao().insertTripMessageOrIgnore(message);
            }
        });
    }*/
}

File: ovikl-ios/Caoutch/TripSingleton.swift
//
//  TripSingleton.swift
//  Caoutch
//
//  Created by Mohamed Dahroug on 8/3/18.
//  Copyright © 2018 Caoutch. All rights reserved.
//

import Foundation
import CoreData

class TripSingelton{
    static var mInstance = TripSingelton();
    var _id:String!;
    var createTime:Date!;
    var updateTime:Date!;
    var state:String!;
    var clientId:String!;
    var driverId:String!;
    var clientLat:Double!;
    var clientLng:Double!;
    var driverLat:Double!;
    var driverLng:Double!;
    var prMin:Double!;
    var prBase:Double!;
    var prKM:Double!;
    var prMinute:Double!;
    /*var prLngKM:Double!;
    var prLngMinute:Double!;
    var lngKM:Double!;*/
    var cancelledBy:String!;
    var cur:String!;
    var distance:Double!;
    var duration:Double!;
    var cost:Double!;
    var lang:String!;
    var dateFormatter=DateFormatter();
    var msgId:String!
    var promoPercentage:Double!;
    var maxPromoAmount:Double!;
    var paymentMethod:String!;
    var stripeLast4:String!;
    var stripeCustomerId:String!;
    var appDelegate:AppDelegate!
    private init(){
        lang=Locale.preferredLanguages[0];
        dateFormatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'";
        dateFormatter.timeZone = TimeZone(identifier: "GMT");
        dateFormatter.locale = Locale(identifier: "en_US_POSIX")
    }
    
    public func saveTrip(managedContext:NSManagedObjectContext){
        print("func saveTrip")
        let tripEntity = NSEntityDescription.entity(forEntityName: "Trip", in: managedContext)!
        let fetchRequest = NSFetchRequest<NSFetchRequestResult>(entityName: "Trip")
        let trip:NSManagedObject ;
        fetchRequest.predicate = NSPredicate(format: "tripId = %@ AND state = %@",argumentArray: [_id, state])
        
        do {
            let results = try managedContext.fetch(fetchRequest) as? [NSManagedObject]
            if results?.count != 0{
                if let trip2=results?.first{
                    trip=trip2;
                }
                else{
                    trip = NSManagedObject(entity: tripEntity, insertInto: managedContext);
                }
            }
            else{
                trip = NSManagedObject(entity: tripEntity, insertInto: managedContext);
            }
        } catch {
            print("Fetch Failed: \(error)");
            trip = NSManagedObject(entity: tripEntity, insertInto: managedContext);
        }
        
        
        trip.setValue(_id, forKeyPath: "tripId");
        trip.setValue(createTime, forKeyPath: "createTime");
        trip.setValue(updateTime, forKeyPath: "updateTime");
        trip.setValue(state, forKeyPath: "state");
        trip.setValue(clientId, forKeyPath: "clientId");
        trip.setValue(driverId, forKeyPath: "driverId");
        trip.setValue(clientLat, forKeyPath: "clientLat");
        trip.setValue(clientLng, forKeyPath: "clientLng");
        trip.setValue(driverLat, forKeyPath: "driverLat");
        trip.setValue(driverLng, forKeyPath: "driverLng");
        trip.setValue(prMin, forKeyPath: "prMin");
        trip.setValue(prBase, forKeyPath: "prBase");
        trip.setValue(prKM, forKeyPath: "prKM");
        trip.setValue(prMinute, forKeyPath: "prMinute");
        /*trip.setValue(prLngKM, forKeyPath: "prLngKM");
        trip.setValue(prLngMinute, forKeyPath: "prLngMinute");
        trip.setValue(lngKM, forKeyPath: "lngKM");*/
        trip.setValue(cancelledBy, forKeyPath: "cancelledBy");
        trip.setValue(cur, forKeyPath: "cur");
        trip.setValue(distance, forKeyPath: "distance");
        trip.setValue(duration, forKeyPath: "duration");
        trip.setValue(cost, forKeyPath: "cost");
        trip.setValue(msgId, forKeyPath: "msgId");
        trip.setValue(promoPercentage, forKeyPath: "promoPercentage");
        trip.setValue(maxPromoAmount, forKeyPath: "maxPromoAmount");
        trip.setValue(paymentMethod, forKeyPath: "paymentMethod");
        trip.setValue(stripeLast4, forKeyPath: "stripeLast4");
        trip.setValue(stripeCustomerId, forKeyPath: "stripeCustomerId");
        do {
            try managedContext.save()
            print("trip \(_id) \(state) is saved")
        } catch let error as NSError {
            print("Could not save. \(error), \(error.userInfo)")
        }
    
    }
    
    public func setTrip(trip:[String:Any]){
        if let _id=trip["_id"] as? String{self._id=_id};
        if let createTimeStr=trip["createTime"] as? String,let createTime=dateFormatter.date(from: createTimeStr) {self.createTime=createTime};
        if let updateTimeStr=trip["updateTime"] as? String,let updateTime=dateFormatter.date(from: updateTimeStr) {self.updateTime=updateTime};
        if let state=trip["state"] as? String{self.state=state};
        if let clientId=trip["clientId"] as? String{self.clientId=clientId};
        if let driverId=trip["driverId"] as? String{self.driverId=driverId};
        if let clientLat=trip["clientLat"] as? String{self.clientLat=Double(clientLat)};
        if let clientLng=trip["clientLng"] as? String{self.clientLng=Double(clientLng)};
        if let driverLat=trip["driverLat"] as? String{self.driverLat=Double(driverLat)};
        if let driverLng=trip["driverLng"] as? String{self.driverLng=Double(driverLng)};
        if let prMin=trip["prMin"] as? Double{self.prMin=prMin};
        if let prBase=trip["prBase"] as? Double{self.prBase=prBase};
        if let prKM=trip["prKM"] as? Double{self.prKM=prKM};
        if let prMinute=trip["prMinute"] as? Double{self.prMinute=prMinute};
//        if let prLngKM=trip["prLngKM"] as? Double{self.prLngKM=prLngKM};
//        if let prLngMinute=trip["prLngMinute"] as? Double{self.prLngMinute=prLngMinute};
//        if let lngKM=trip["lngKM"] as? Double{self.lngKM=lngKM};
        if let cancelledBy=trip["cancelledBy"] as? String{self.cancelledBy=cancelledBy};
        
        if let distance=trip["distance"] as? Double{self.distance=distance};
        if let duration=trip["duration"] as? Double{self.duration=duration};
        if let cost=trip["cost"] as? Double{self.cost=cost};
        if let msgId=trip["msgId"] as? String{self.msgId=msgId};
        
        if let cur=trip[lang+"cur"] as? String{
            self.cur=cur
        }
        else if let cur=trip["cur"] as? String{
            self.cur=cur
        }
        if let promoPercentage=trip["promoPercentage"] as? Double{self.promoPercentage=promoPercentage};
        if let maxPromoAmount=trip["maxPromoAmount"] as? Double{self.maxPromoAmount=maxPromoAmount};
        if let paymentMethod=trip["paymentMethod"] as? String{self.paymentMethod=paymentMethod};
        if let stripeLast4=trip["stripeLast4"] as? String{self.stripeLast4=stripeLast4};
        if let stripeCustomerId=trip["stripeCustomerId"] as? String{self.stripeCustomerId=stripeCustomerId};
    }
    
    public func setObject(trip:NSManagedObject){
        if let _id=trip.value(forKey:"tripId") as? String{self._id=_id};
        if let createTime=trip.value(forKey:"createTime") as? Date {self.createTime=createTime};
        if let updateTime=trip.value(forKey:"updateTime") as? Date {self.updateTime=updateTime};
        if let state=trip.value(forKey:"state") as? String{self.state=state};
        if let clientId=trip.value(forKey:"clientId") as? String{self.clientId=clientId};
        if let driverId=trip.value(forKey:"driverId") as? String{self.driverId=driverId};
        if let clientLat=trip.value(forKey:"clientLat") as? Double{self.clientLat=clientLat};
        if let clientLng=trip.value(forKey:"clientLng") as? Double{self.clientLng=clientLng};
        if let driverLat=trip.value(forKey:"driverLat") as? Double{self.driverLat=driverLat};
        if let driverLng=trip.value(forKey:"driverLng") as? Double{self.driverLng=driverLng};
        if let prMin=trip.value(forKey:"prMin") as? Double{self.prMin=prMin};
        if let prBase=trip.value(forKey:"prBase") as? Double{self.prBase=prBase};
        if let prKM=trip.value(forKey:"prKM") as? Double{self.prKM=prKM};
        if let prMinute=trip.value(forKey:"prMinute") as? Double{self.prMinute=prMinute};
//        if let prLngKM=trip.value(forKey:"prLngKM") as? Double{self.prLngKM=prLngKM};
//        if let prLngMinute=trip.value(forKey:"prLngMinute") as? Double{self.prLngMinute=prLngMinute};
//        if let lngKM=trip.value(forKey:"lngKM") as? Double{self.lngKM=lngKM};
        if let cancelledBy=trip.value(forKey:"cancelledBy") as? String{self.cancelledBy=cancelledBy};
        if let cur=trip.value(forKey:"cur") as? String{self.cur=cur};
        if let distance=trip.value(forKey:"distance") as? Double{self.distance=distance};
        if let duration=trip.value(forKey:"duration") as? Double{self.duration=duration};
        if let cost=trip.value(forKey:"cost") as? Double{self.cost=cost};
        if let msgId=trip.value(forKey:"msgId") as? String{self.msgId=msgId};
        if let promoPercentage=trip.value(forKey:"promoPercentage") as? Double{self.promoPercentage=promoPercentage};
        if let maxPromoAmount=trip.value(forKey:"maxPromoAmount") as? Double{self.maxPromoAmount=maxPromoAmount};
        if let paymentMethod=trip.value(forKey:"paymentMethod") as? String{self.paymentMethod=paymentMethod};
        if let stripeLast4=trip.value(forKey:"stripeLast4") as? String{self.stripeLast4=stripeLast4};
        if let stripeCustomerId=trip.value(forKey:"stripeCustomerId") as? String{self.stripeCustomerId=stripeCustomerId};
    }
    
    
    public func setResponseTrip(trip:TCPResponse.Trip){
        if trip._id != nil{
            _id=trip._id
        }
        if trip.createTime != nil{
            createTime=trip.createTime
        }
        if trip.updateTime != nil{
            updateTime=trip.updateTime
        }
        if trip.state != nil{
            state=trip.state
        }
        if trip.clientId != nil{
            clientId=trip.clientId
        }
        if trip.driverId != nil{
            driverId=trip.driverId
        }
        if trip.clientLat != nil{
            clientLat=trip.clientLat
        }
        if trip.clientLng != nil{
            clientLng=trip.clientLng
        }
        if trip.driverLat != nil{
            driverLat=trip.driverLat
        }
        if trip.driverLng != nil{
            driverLng=trip.driverLng
        }
        if trip.prMin != nil{
            prMin=trip.prMin
        }
        if trip.prBase != nil{
            prBase=trip.prBase
        }
        if trip.prKM != nil{
            prKM=trip.prKM
        }
        if trip.prMinute != nil{
            prMinute=trip.prMinute
        }
//        prLngKM=trip.prLngKM
//        prLngMinute=trip.prLngMinute
//        lngKM=trip.lngKM
        if trip.cancelledBy != nil{
            cancelledBy=trip.cancelledBy
        }
        if trip.cur != nil{
            cur=trip.cur
        }
        if trip.distance != nil{
            distance=trip.distance
        }
        if trip.duration != nil{
            duration=trip.duration
        }
        if trip.cost != nil{
            cost=trip.cost
        }
        if trip.msgId != nil{
            msgId=trip.msgId
        }
    }
    
    
    
    
    
    public func getMap()->[String: Any]{
        var parameters: [String: Any]=[:];
        
        parameters["type"]=appDelegate.user.type
        parameters["auth_id"]=appDelegate.user._id
        parameters["time"]=Constants.dateFormatter.string(from: Date());
        parameters["msgId"]=UUID().uuidString;
        parameters["lang"] = Locale.current.languageCode
        parameters["ver"] = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String
        
        if let _=_id {parameters["_id"]=_id;}
        //if let _=createTime {parameters["createTime"]=dateFormatter.string(from: createTime);}
        //if let _=updateTime {parameters["updateTime"]=dateFormatter.string(from: updateTime);}
        if let _=state {parameters["state"]=state;}
        if let _=clientId {parameters["clientId"]=clientId;}
        if let _=driverId {parameters["driverId"]=driverId;}
        if let _=clientLat {parameters["clientLat"]=String(clientLat);}
        if let _=clientLng {parameters["clientLng"]=String(clientLng);}
        if let _=driverLat {parameters["driverLat"]=String(driverLat);}
        if let _=driverLng {parameters["driverLng"]=String(driverLng);}
        /*
        if let _=prMin {parameters["prMin"]=prMin;}
        if let _=prBase {parameters["prBase"]=prBase;}
        if let _=prKM {parameters["prKM"]=prKM;}
        if let _=prMinute {parameters["prMinute"]=prMinute;}
        if let _=prLngKM {parameters["prLngKM"]=prLngKM;}
        if let _=prLngMinute {parameters["prLngMinute"]=prLngMinute;}
        if let _=lngKM {parameters["lngKM"]=lngKM;}*/
        if let _=cur {parameters["cur"]=cur;}
         
        if let _=cancelledBy {parameters["cancelledBy"]=cancelledBy;}
        if let _=distance {parameters["distance"]=distance;}
        if let _=duration {parameters["duration"]=duration;}
        if let _=cost {parameters["cost"]=cost;}
//        if let _=promoPercentage {parameters["promoPercentage"]=promoPercentage;}
//        if let _=maxPromoAmount {parameters["maxPromoAmount"]=maxPromoAmount;}
//        if let _=paymentMethod {parameters["paymentMethod"]=paymentMethod;}
//        if let _=stripeLast4 {parameters["stripeLast4"]=stripeLast4;}
//        if let _=stripeCustomerId {parameters["stripeCustomerId"]=stripeCustomerId;}
        return parameters
    }
    
    public func reset(){
        
        _id=nil;
        createTime=nil;
        updateTime=nil;
        //startTime=nil;
        //endTime=nil;
        state=nil;
        clientId=nil;
        driverId=nil;
        clientLat=nil;
        clientLng=nil;
        driverLat=nil;
        driverLng=nil;
        prMin=nil;
        prBase=nil;
        prKM=nil;
        prMinute=nil;
//        prLngKM=nil;
//        prLngMinute=nil;
//        lngKM=nil;
        cur=nil;
        cancelledBy=nil;
        distance=nil;
        duration=nil;
        cost=nil;
        msgId=nil;
        promoPercentage=nil;
        maxPromoAmount=nil;
        paymentMethod=nil;
        stripeLast4=nil;
        stripeCustomerId=nil;
    }
}
